### DISCUSSION TITLE
Pydantic serialization blocking synchronous endpoints with global capacity limiter

### DISCUSSION THREAD

**Discussion by guillaumegenthial** [fastapi/fastapi#11401] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
############################################################
# app.py, run with uvicorn app:app --reload

import time
from datetime import datetime, timezone

from anyio import CapacityLimiter
from anyio.lowlevel import RunVar
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class SomeModel(BaseModel):
    x: int = 0


@app.on_event("startup")
def startup():
    RunVar("_default_thread_limiter").set(CapacityLimiter(1))


@app.get("/sync-with-model", response_model=SomeModel)
def sync_with_model(request_no):
    start = datetime.now(timezone.utc)
    print("start", request_no, start)
    time.sleep(1)
    end = datetime.now(timezone.utc)
    print("end", request_no, end)
    return SomeModel(x=request_no)


@app.get("/sync-without-model")
def sync_without_model(request_no):
    start = datetime.now(timezone.utc)
    print("start", request_no, start)
    time.sleep(1)
    end = datetime.now(timezone.utc)
    print("end", request_no, end)
    return SomeModel(x=request_no)

############################################################
# client.py -- run with python client.py

from datetime import datetime, timezone
from multiprocessing import Process

import requests


def send_req(request_no):
    print("start", request_no, datetime.now(timezone.utc))
    requests.get(f"http://localhost:8000/sync-with-model?request_no={request_no}")
    print("end", request_no, datetime.now(timezone.utc))


if __name__ == "__main__":
    num_requests = 6
    processes = [Process(target=send_req, args=(i,)) for i in range(num_requests)]
    for i in range(num_requests):
        processes[i].start()
        print(f"Started process {i}")
```

## Context & Motivation

TL:DR: reduce synchronous endpoints pool size

I have a Fast API app that uses synchronous endpoints like this

```python
@app.get("/some-route", response_model=SomeModel)
def some_route(some_params):
    # Connect to DB
    db = get_new_db_connection(pool_size=10)
    return SomeModel(...)
```

Fast API runs those endpoints using Starlette's `run_in_threadpool` function.

By default, it uses a threadpool of 40 threads (see [`anyio` documentation](https://anyio.readthedocs.io/en/stable/threads.html#adjusting-the-default-maximum-worker-thread-count)).

My app also needs to connect to some database using a pool of connections. This pool size **needs** to be smaller than 40 (say 10), so I need to change the default thread pool size. 

Otherwise when I issue more than 10 concurrent requests, the line 

```python
db = get_new_db_connection(pool_size=10)
```

will raise an error saying "I've exhausted my number of  connections"

## Current workaround

I've seen this issue https://github.com/tiangolo/fastapi/issues/4221 that suggests doing this

```python
@app.on_event("startup")
def startup():
    RunVar("_default_thread_limiter").set(CapacityLimiter(10))
```

This works well to limit the number of active threads for dependencies + endpoint code.

## Problem with current workaround

**However**, I've experienced some blocking behavior at serialization. 

Instead of returning one after the other, concurrent requests are blocked from being sent back to the client and instead wait together.

To reproduce, see the example code. Summary:
1. Sets a pool size limiter to `1`
2. Defines a route **with a `response_model`** (if you remove this, it works fine) that waits for 1 second
3. Issues 6 concurrent requests to the router

Result:
a. Nothing for 6 seconds, then all the responses at once **(blocking)**

When you remove the response model you get
b. One response every second (**expected behavior**)

## Suspected root cause

The synchronous endpoints serialization step seems to call `run_in_threadpool` as well if the endpoint is synchronous. See

https://github.com/tiangolo/fastapi/blob/9490491595aa2f9eb0326109237be762dfff06a5/fastapi/routing.py#L144-L149

As it doesn't use a dedicated limiter, it is limited by the `_default_thread_limiter` set at app startup.

This means that when concurrent requests reach the serialization steps, they block and wait for all the requests to be processed first, and only then can they start sending data back.

It seems that a similar issue was found + fixed for dependencies, see https://github.com/tiangolo/fastapi/pull/5122 by using a custom limiter.

I'm not familiar enough with the internals to make a change myself but it'd probably be a good idea to apply the same non-blocking mechanism as dependencies to serialization.

## Related 

https://github.com/tiangolo/fastapi/issues/4221

https://github.com/tiangolo/fastapi/discussions/6347

https://github.com/encode/starlette/discussions/1690

https://github.com/tiangolo/fastapi/pull/5122

https://news.ycombinator.com/item?id=25993229

https://github.com/tiangolo/fastapi/issues/3205 (on managing DB connections with Depends / FastAPI)

### Operating System

macOS

### FastAPI Version

0.110.1

### Pydantic Version

2.6.4

### Python Version

3.11.8

**Comment by guillaumegenthial** [comment]
(This discussion should probably be an issue but it seems the process is to open a question first, then wait for confirmation that it's an issue)

**Comment by guillaumegenthial** [comment]
For anyone reading this, I ended up embracing ASGI and only using `async` endpoints to avoid this kind of situation.

However, **make sure that all your dependencies are async** otherwise you will run into some issues.

If one of the dependencies in the stack is not async, it first opens ALL the connections to the DB then executes the endpoints one by one. (the dependencies are executed in a threadpool separately)

See the code below

```python
# app.py
import time
from contextlib import contextmanager
from datetime import datetime, timezone

from fastapi import Depends, FastAPI
from pydantic import BaseModel

app = FastAPI()


class SomeModel(BaseModel):
    x: int = 0


_NUM_CONNECTIONS = 0


@contextmanager
def _get_db():
    global _NUM_CONNECTIONS
    _NUM_CONNECTIONS += 1
    print(f"------ open db {_NUM_CONNECTIONS}")
    yield _NUM_CONNECTIONS
    print(f"------ close db {_NUM_CONNECTIONS}")
    _NUM_CONNECTIONS -= 1


async def get_db():  # If you change this to sync, opens all the DB connections at once
    with _get_db() as db:
        yield db


@app.get("/async-with-model", response_model=SomeModel)
async def async_with_model(request_no, db=Depends(get_db)):
    start = datetime.now(timezone.utc)
    print("start", request_no, start)
    time.sleep(1)
    end = datetime.now(timezone.utc)
    print("end", request_no, end)
    return SomeModel(x=request_no)


# client.py
from datetime import datetime, timezone
from multiprocessing import Process

import requests


def send_req(request_no):
    print("start", request_no, datetime.now(timezone.utc))
    requests.get(f"http://localhost:8000/async-with-model?request_no={request_no}")
    print("end", request_no, datetime.now(timezone.utc))


if __name__ == "__main__":
    num_requests = 6
    processes = [Process(target=send_req, args=(i,)) for i in range(num_requests)]
    for i in range(num_requests):
        processes[i].start()
        print(f"Started process {i}")

```

**Reply by YuriiMotov** [reply]
Since you use `time.sleep(1)` inside the async endpoint, your event loop will be blocked, won't it?

**Reply by guillaumegenthial** [reply]
@YuriiMotov yes, that's on purpose.

It just explains a way to implement "one request at a time".

It uses the blocking time to illustrate the impact of async / non-async dependencies.

NOTE: this approach uses concurrency to avoid creating too many DB connections. But it doesn't solve the parallelism issue with serialization that is the real topic of the issue BTW.

**Comment by guillaumegenthial** [comment]
A thought after reading https://github.com/tiangolo/fastapi/issues/3205.

It seems that for a synchronous route with a synchronous dependency + serialization like so

```python
@app.get("/some-route", response_model=SomeModel)
def some_route(db=Depends(get_db)):
    # Do some stuff
```

There are 3 tasks that are **independently** queued to the threadpool
a. The dependency (`Depends(get_db)`)
b. The route code itself (`# Do some stuff`)
c. The serialization (ultimately added to the route because of `response_model`)

When sending `n` concurrent requests, the order of tasks sent to the threadpool **is not**
1a -> 1b -> 1c -> 2a -> 2b -> 2c -> etc.

But something on a **first-come / first-serve** basis. Typically
1a -> 1b -> 2a -> 3a -> 4a -> etc.

This causes unexpected behaviors and deadlocks since a completed request may need to wait for other requests to complete before being able to serialize and send the response back to the client for example.

A specific case of this issue for context dependencies closure was solved in https://github.com/tiangolo/fastapi/pull/5122.

I wonder if a different design choice in how fastapi executes dependencies / serialization for synchronous routes might make more sense without breaking anything. 

**Would it be possible to make sure that the (a, b, c) steps are executed together, ideally in the same thread ?**

In other words, instead of doing (pseudo code)

```python
for request in requests:
    dependencies = await run_in_threadpool(get_dependencies, request)
    result = await run_in_threadpool(execute_route, request, dependencies)
    serialized = await run_in_threadpool(serialize_result, result)
    await send_to_client(serialized)
```

execute by group in the same thread

```python
def execute_and_send(request)
    dependencies = get_dependencies(request)
    result = execute_route(request, dependencies)
    serialized = serialize_result(result)
    send_to_client(serialized)

for request in requests:
    await run_in_threadpool(execute_and_send, request)
```

**Reply by YuriiMotov** [reply, marked as answer]
> execute by group in the same thread

I think this is not possible. Dependencies and route function are not always sync. So, you can't just combine them into one sync function.

I think the solution might be to  prioritize tasks in threadpool:
- highest priority for response validation tasks
- middle priority for endpoint execution tasks
- lowest priority for dependencies

But I'm not sure it's needed.

In your case you can limit the number of concurrently used DB sessions by using `asyncio.Semaphore`:

```
import asyncio
import time
from datetime import datetime, timezone
from typing import Annotated, cast

from fastapi import Depends, FastAPI, Request


CONNECTION_LIMIT = 2
connection_counter = {"acquired": 0}


async def lifespan(app: FastAPI):
    conn_limit_semaphore = asyncio.Semaphore(CONNECTION_LIMIT)  # Limit to 10 concurrent conn.
    yield {"connection_limit_sem": conn_limit_semaphore}


app = FastAPI(lifespan=lifespan)


async def get_new_db_connection(request: Request):
    sem = cast(asyncio.Semaphore, request.state.connection_limit_sem)
    async with sem:
        connection_counter["acquired"] += 1
        print(f"Connection acquired (total: {connection_counter['acquired']})")
        yield "connection"
        connection_counter["acquired"] -= 1
        print(f"Connection released (total: {connection_counter['acquired']})")


@app.get("/test")
def index(request_no: int, conn: Annotated[str, Depends(get_new_db_connection)]):
    start = datetime.now(timezone.utc)
    print("start", request_no, start)
    time.sleep(1)
    end = datetime.now(timezone.utc)
    print("end", request_no, end)
    return "Hello World"

```
